sotodlib/tod_ops/gapfill.py

```diff
@@ -0,0 +1,331 @@
+    """This class is a simple container for a list (of length n_dets) of
+    Extract objects of equal length (n_samps), to provide abstract
+    access to data with shape (n_dets, n_samps).  Each child Extract
+    can be accessed with [item] indexing.  Most methods have (tod,
+    signal=None) signature, and are simple loops that call the method
+    of the same name on each child Extract object.
+
+    """
+    def __init__(self, items=None):
+        """The complete list of Extract objects should be passed in."""
+        if items is None:
+            items = []
+        self.items = items
+
+    def __repr__(self):
+        """This repr shows the shape of the full array, along with the
+        fraction (as a percentage) of the full space that is actually
+        tracked.
+
+        """
+        frac = (sum([e.n_ex for e in self.items]) /
+                (self.shape[0] * self.shape[1]))
+        return 'ExtractMatrix(' + ','.join(map(str, self.shape)) + \
+            '@%.1f%%)' % (frac*100)
+
+    def __len__(self):
+        return self.shape[0]
+
+    def copy(self):
+        return ExtractMatrix([x.copy() for x in self.items])
+
+    @property
+    def shape(self):
+        if len(self.items) == 0:
+            return (0,)
+        return (len(self.items),) + self.items[0].shape
+
+    def __getitem__(self, index):
+        return self.items[index]
+
+    def expand(self, fill_value=0):
+        """Expands the extract into a full-size data array, filling missing
+        bits with fill_value."""
+        dtype = self.items[0].dtype
+        signal = np.zeros(self.shape, dtype) + fill_value
+        for o, i in zip(signal, self.items):
+            i.patch(o)
+        return signal
+
+    def swap(self, tod, signal=None):
+        """Swaps the current extract with the tracked samples of full data
+        vector signal."""
+        if signal is None:
+            signal = tod.signal
+        for o, i in zip(signal, self.items):
+            i.swap(o)
+        return signal
+
+    def patch(self, tod, signal=None):
+        """Copies the extract into the signal vector.  Untracked samples are
+        not modified.
+
+        """
+        if signal is None:
+            signal = tod.signal
+        for o, i in zip(signal, self.items):
+            i.patch(o)
+        return signal
+
+    def accumulate(self, tod, signal, scale):
+        """Adds the extract, scaled by factor scale, into the full length
+        signal vector.  Untracked samples are not modified.
+
+        """
+        if signal is None:
+            signal = tod.signal
+        for o, i in zip(signal, self.items):
+            i.accumulate(o, scale)
+        return signal
+
+
+def get_gap_fill_single(data, flags, nbuf=10, order=1, inplace=False):
+    """Computes samples to fill the gaps in data identified by flags.
+    Each flagged segment is modeled with a polynomial of the order
+    specified, based on up to nbuf points on each side of the segment.
+
```

Comment:
General questions about the use case of this gap-filling:  If you apply this kind of gap-filling technique to data which contains signal + noise, it will put some amount of fake signal in the gap which could then leak back into unflagged samples when applying a filter.  The only kind of gap-filling I have experience with is the case of first estimating a noise timestream (by subtracting a signal estimate) and then using that to build a constrained noise realization across all gaps simultaneously (to ensure that long-timescale correlations are included).

Of course if your timestream is completely noise dominated, then you can skip the subtraction of the signal estimate.  So are you just treating atmosphere as "noise" in this case?

Even in this noise dominated regime, are you not concerned about gap-filling individual gaps versus all of them simultaneously for the entire CES?

Reply:
I agree that, because of filtering, filling a single sample potentially has long-range consequences.  The ACT map-maker (IIRC) actually solves for the value of flagged samples, so the gap-filled values are just starting points in the iterative solution.

However, the gap-filled data _are_ used to estimate a noise model for the max-L solution.  As you imply, the thing to worry about is the atmosphere; it is 1/f "noise" with strong correlations between detectors.  We've actually seen that naive gap-filling can lead to noise models that have higher map noise, i.e. are highly sub-optimal, especially when the data contain gaps >= 1/f_knee.  Our modern approach does two-passes -- at first you just fill single time-stream gaps, then you do a PCA to constrain the low-frequency correlated modes and use that for a superior gap fill.  Then you estimate a (correlated) noise model.

But I agree with your statements (especially below) that some care should be taken with mapping of gap-filled time-streams.  The present routines will get us through basic troubles such as "I can't plot or fft this timestream because there are huge glitches", and similar.